File: hiob_contracts/beat_coverage.py
"""Exact run-level beat coverage and serial fan-in receipts.

The per-beat lane builders are intentionally free to finish out of order.  This
contract is the terminal join: it proves that every expected beat has exactly
one successful receipt for every required lane, all receipts belong to the same
run/workspace/script package/beat plan, and the resulting coverage is immutable
through a canonical digest.
"""
from __future__ import annotations

from dataclasses import asdict, dataclass, fields, replace
from typing import Any

from .factory.digest import is_digest, sha256_digest


BEAT_COVERAGE_CONTRACT_VERSION_V1 = "BeatCoverage.v1"
SERIAL_FAN_IN_RECEIPT_CONTRACT_VERSION_V1 = "SerialFanInReceipt.v1"
BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1 = "BeatLaneTerminalReceipt.v1"

# The serial F1 spine requires these lanes.  Apollo/SFX is an optional lane and
# may be included in a receipt without changing the required coverage set.
DEFAULT_BEAT_COVERAGE_LANES_V1 = ("athena", "orpheus_vo", "atropos")
TERMINAL_BEAT_LANE_STATUSES_V1 = frozenset(
    {"succeeded", "failed", "cancelled", "blocked", "needs_human"}
)
_COVERAGE_CONTRACT_VERSIONS = frozenset(
    {
        BEAT_COVERAGE_CONTRACT_VERSION_V1,
        SERIAL_FAN_IN_RECEIPT_CONTRACT_VERSION_V1,
    }
)


def _nonblank(value: Any) -> bool:
    return bool(str(value or "").strip())


@dataclass(frozen=True)
class BeatLaneTerminalReceiptV1:
    """One terminal result for one beat/lane pair."""

    run_id: str
    workspace_id: str
    package_digest: str
    plan_digest: str
    beat_index: int
    lane: str
    status: str = "succeeded"
    output_digest: str | None = None
    receipt_digest: str = ""
    contract_version: str = BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1

    def digest_payload(self) -> dict[str, Any]:
        payload = self.to_dict()
        payload.pop("receipt_digest", None)
        return payload

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, value: dict[str, Any]) -> "BeatLaneTerminalReceiptV1":
        accepted = {item.name for item in fields(cls)}
        return cls(**{key: item for key, item in value.items() if key in accepted})

    @classmethod
    def create(cls, **values: Any) -> "BeatLaneTerminalReceiptV1":
        draft = cls(**{**values, "receipt_digest": ""})
        return replace(draft, receipt_digest=sha256_digest(draft.digest_payload()))

    def validate(self) -> list[str]:
        errors: list[str] = []
        for name in ("run_id", "workspace_id", "lane"):
            if not _nonblank(getattr(self, name)):
                errors.append(f"{name} is required")
        if self.contract_version != BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1:
            errors.append("unsupported lane receipt contract_version")
        if isinstance(self.beat_index, bool) or not isinstance(self.beat_index, int):
            errors.append("beat_index must be an integer")
        elif self.beat_index < 0:
            errors.append("beat_index must be non-negative")
        for name in ("package_digest", "plan_digest"):
            if not is_digest(getattr(self, name)):
                errors.append(f"{name} must be a sha256 digest")
        if self.output_digest is not None and not is_digest(self.output_digest):
            errors.append("output_digest must be a sha256 digest when present")
        if self.status not in TERMINAL_BEAT_LANE_STATUSES_V1:
            errors.append(f"status is not terminal: {self.status!r}")
        elif self.status != "succeeded":
            errors.append("lane receipt status must be succeeded")
        if not is_digest(self.receipt_digest):
            errors.append("receipt_digest must be a sha256 digest")
        elif self.receipt_digest != sha256_digest(self.digest_payload()):
            errors.append("receipt_digest does not match lane receipt payload")
        return errors


@dataclass(frozen=True)
class BeatCoverageV1:
    """Terminal exact beat coverage for a serial fan-in run."""

    run_id: str
    workspace_id: str
    package_digest: str
    plan_digest: str
    expected_n_beats: int
    expected_beat_indices: tuple[int, ...]
    lane_receipts: tuple[BeatLaneTerminalReceiptV1, ...]
    required_lanes: tuple[str, ...] = DEFAULT_BEAT_COVERAGE_LANES_V1
    coverage_digest: str = ""
    contract_version: str = BEAT_COVERAGE_CONTRACT_VERSION_V1

    def digest_payload(self) -> dict[str, Any]:
        payload = self.to_dict()
        payload.pop("coverage_digest", None)
        # Completion order is intentionally not part of run identity.  Sort
        # the exact lane set before hashing so workers may finish out of order.
        payload["lane_receipts"] = sorted(
            payload["lane_receipts"],
            key=lambda receipt: (int(receipt["beat_index"]), str(receipt["lane"])),
        )
        return payload

    def to_dict(self) -> dict[str, Any]:
        return {
            "run_id": self.run_id,
            "workspace_id": self.workspace_id,
            "package_digest": self.package_digest,
            "plan_digest": self.plan_digest,
            "expected_n_beats": self.expected_n_beats,
            "expected_beat_indices": list(self.expected_beat_indices),
            "lane_receipts": [receipt.to_dict() for receipt in self.lane_receipts],
            "required_lanes": list(self.required_lanes),
            "coverage_digest": self.coverage_digest,
            "contract_version": self.contract_version,
        }

    @classmethod
    def from_dict(cls, value: dict[str, Any]) -> "BeatCoverageV1":
        receipts = value.get("lane_receipts")
        if receipts is None:
            receipts = value.get("terminal_receipts") or value.get("receipts") or ()
        package_digest = value.get("package_digest") or value.get("script_package_digest") or ""
        plan_digest = value.get("plan_digest") or value.get("beat_plan_digest") or ""
        return cls(
            run_id=str(value.get("run_id") or ""),
            workspace_id=str(value.get("workspace_id") or ""),
            package_digest=str(package_digest),
            plan_digest=str(plan_digest),
            expected_n_beats=int(value.get("expected_n_beats") or 0),
            expected_beat_indices=tuple(value.get("expected_beat_indices") or ()),
            lane_receipts=tuple(
                receipt
                if isinstance(receipt, BeatLaneTerminalReceiptV1)
                else BeatLaneTerminalReceiptV1.from_dict(receipt)
                for receipt in (receipts or ())
            ),
            required_lanes=tuple(
                value.get("required_lanes") or DEFAULT_BEAT_COVERAGE_LANES_V1
            ),
            coverage_digest=str(value.get("coverage_digest") or ""),
            contract_version=str(
                value.get("contract_version") or BEAT_COVERAGE_CONTRACT_VERSION_V1
            ),
        )

    @classmethod
    def create(cls, **values: Any) -> "BeatCoverageV1":
        raw_receipts = values.get("lane_receipts") or ()
        receipts = tuple(
            receipt
            if isinstance(receipt, BeatLaneTerminalReceiptV1)
            else BeatLaneTerminalReceiptV1.from_dict(receipt)
            for receipt in raw_receipts
        )
        receipts = tuple(
            receipt
            if receipt.receipt_digest
            else BeatLaneTerminalReceiptV1.create(**receipt.to_dict())
            for receipt in receipts
        )
        draft = cls(**{**values, "lane_receipts": receipts, "coverage_digest": ""})
        return replace(draft, coverage_digest=sha256_digest(draft.digest_payload()))

    def validate(self) -> list[str]:
        errors: list[str] = []
        for name in ("run_id", "workspace_id"):
            if not _nonblank(getattr(self, name)):
                errors.append(f"{name} is required")
        if self.contract_version not in _COVERAGE_CONTRACT_VERSIONS:
            errors.append("unsupported coverage contract_version")
        for name in ("package_digest", "plan_digest"):
            if not is_digest(getattr(self, name)):
                errors.append(f"{name} must be a sha256 digest")
        if (
            isinstance(self.expected_n_beats, bool)
            or not isinstance(self.expected_n_beats, int)
            or self.expected_n_beats < 1
        ):
            errors.append("expected_n_beats must be a positive integer")
        valid_n = (
            self.expected_n_beats
            if isinstance(self.expected_n_beats, int)
            and not isinstance(self.expected_n_beats, bool)
            and self.expected_n_beats > 0
            else 0
        )
        expected = list(range(valid_n))
        observed_indices = list(self.expected_beat_indices)
        if observed_indices != expected:
            errors.append(
                "expected_beat_indices must be exactly 0..N-1 without duplicates, "
                "holes, or out-of-range values"
            )

        required = tuple(self.required_lanes)
        if not required or any(not _nonblank(lane) for lane in required):
            errors.append("required_lanes must contain non-blank lane names")
        if len(required) != len(set(required)):
            errors.append("required_lanes contains duplicate lanes")
        for required_lane in DEFAULT_BEAT_COVERAGE_LANES_V1:
            if required_lane not in required:
                errors.append(f"required_lanes must include {required_lane}")

        seen: set[tuple[int, str]] = set()
        for receipt in self.lane_receipts:
            if not isinstance(receipt, BeatLaneTerminalReceiptV1):
                errors.append("lane_receipts must contain BeatLaneTerminalReceiptV1")
                continue
            errors.extend(f"lane receipt: {error}" for error in receipt.validate())
            key = (receipt.beat_index, receipt.lane)
            if key in seen:
                errors.append(f"duplicate lane receipt for beat {key[0]} lane {key[1]!r}")
            seen.add(key)
            if receipt.run_id != self.run_id:
                errors.append("lane receipt run_id does not match coverage")
            if receipt.workspace_id != self.workspace_id:
                errors.append("lane receipt workspace_id does not match coverage")
            if receipt.package_digest != self.package_digest:
                errors.append("lane receipt package_digest does not match coverage")
            if receipt.plan_digest != self.plan_digest:
                errors.append("lane receipt plan_digest does not match coverage")
            if not isinstance(receipt.beat_index, int) or isinstance(receipt.beat_index, bool):
                continue
            if receipt.beat_index < 0 or receipt.beat_index >= valid_n:
                errors.append(f"lane receipt beat_index {receipt.beat_index} out of range")

        expected_pairs = {
            (beat_index, lane)
            for beat_index in expected
            for lane in required
        }
        missing = sorted(expected_pairs - seen)
        if missing:
            errors.append(f"missing lane receipts: {missing}")
        if len(seen) != len(self.lane_receipts):
            # Keep the duplicate diagnostic above, but ensure malformed entries
            # cannot make the terminal receipt appear complete.
            errors.append("lane receipt coverage contains duplicate keys")

        if not is_digest(self.coverage_digest):
            errors.append("coverage_digest must be a sha256 digest")
        elif self.coverage_digest != sha256_digest(self.digest_payload()):
            errors.append("coverage_digest does not match coverage payload")
        return errors

    def assert_valid(self) -> "BeatCoverageV1":
        errors = self.validate()
        if errors:
            raise ValueError("BeatCoverageV1 invalid: " + "; ".join(errors))
        return self


# The two names describe the same terminal payload.  Keeping both names lets
# Star call the receipt by its fan-in role while Contracts callers use the
# shorter coverage name, without creating two divergent digest formats.
SerialFanInReceiptV1 = BeatCoverageV1
BeatTerminalReceiptV1 = BeatLaneTerminalReceiptV1
SerialFanInLaneReceiptV1 = BeatLaneTerminalReceiptV1
LaneTerminalReceiptV1 = BeatLaneTerminalReceiptV1

create_serial_fan_in_receipt_v1 = BeatCoverageV1.create
build_beat_coverage_v1 = BeatCoverageV1.create
build_serial_fan_in_receipt_v1 = BeatCoverageV1.create


__all__ = [
    "BEAT_COVERAGE_CONTRACT_VERSION_V1",
    "SERIAL_FAN_IN_RECEIPT_CONTRACT_VERSION_V1",
    "BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1",
    "DEFAULT_BEAT_COVERAGE_LANES_V1",
    "TERMINAL_BEAT_LANE_STATUSES_V1",
    "BeatLaneTerminalReceiptV1",
    "BeatTerminalReceiptV1",
    "SerialFanInLaneReceiptV1",
    "LaneTerminalReceiptV1",
    "BeatCoverageV1",
    "SerialFanInReceiptV1",
    "create_serial_fan_in_receipt_v1",
    "build_beat_coverage_v1",
    "build_serial_fan_in_receipt_v1",
]

File: ts/src/beat-coverage.ts
/**
 * Exact run-level beat coverage and serial fan-in receipt (Python mirror:
 * hiob_contracts/beat_coverage.py).
 */
import { z } from 'zod';

import { sha256Digest } from './factory/digest.js';

export const BEAT_COVERAGE_CONTRACT_VERSION_V1 = 'BeatCoverage.v1' as const;
export const SERIAL_FAN_IN_RECEIPT_CONTRACT_VERSION_V1 = 'SerialFanInReceipt.v1' as const;
export const BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1 = 'BeatLaneTerminalReceipt.v1' as const;
export const MAX_BEAT_COVERAGE_BEATS_V1 = 16;
export const DEFAULT_BEAT_COVERAGE_LANES_V1 = ['athena', 'orpheus_vo', 'atropos'] as const;
export const TERMINAL_BEAT_LANE_STATUSES_V1 = [
  'succeeded',
  'failed',
  'cancelled',
  'blocked',
  'needs_human',
] as const;

const DigestSchema = z.string().regex(/^sha256:[0-9a-f]{64}$/, 'digest must be sha256:<64 lowercase hex>');
const NonBlankString = z.string().refine(
  (value) => value.trim().length > 0,
  'string must not be blank',
);
const PositiveInteger = z.number().int().positive().max(MAX_BEAT_COVERAGE_BEATS_V1).safe();

export const BeatLaneTerminalReceiptV1Schema = z.object({
  run_id: NonBlankString,
  workspace_id: NonBlankString,
  package_digest: DigestSchema,
  plan_digest: DigestSchema,
  beat_index: z.number().int().nonnegative().safe(),
  lane: NonBlankString,
  status: z.enum(TERMINAL_BEAT_LANE_STATUSES_V1).default('succeeded'),
  output_digest: DigestSchema.nullable().default(null),
  receipt_digest: DigestSchema,
  contract_version: z.literal(BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1)
    .default(BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1),
}).strict().superRefine((receipt, ctx) => {
  if (receipt.status !== 'succeeded') {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['status'], message: 'lane receipt status must be succeeded' });
  }
  const payload = { ...receipt };
  delete (payload as { receipt_digest?: string }).receipt_digest;
  if (receipt.receipt_digest !== sha256Digest(payload)) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['receipt_digest'], message: 'receipt_digest does not match lane receipt payload' });
  }
});

export type BeatLaneTerminalReceiptV1 = z.infer<typeof BeatLaneTerminalReceiptV1Schema>;
export type BeatTerminalReceiptV1 = BeatLaneTerminalReceiptV1;
export type SerialFanInLaneReceiptV1 = BeatLaneTerminalReceiptV1;
export type LaneTerminalReceiptV1 = BeatLaneTerminalReceiptV1;

export type BeatLaneTerminalReceiptV1Input = Pick<
  BeatLaneTerminalReceiptV1,
  'run_id' | 'workspace_id' | 'package_digest' | 'plan_digest' | 'beat_index' | 'lane'
> & Partial<Pick<BeatLaneTerminalReceiptV1, 'status' | 'output_digest' | 'contract_version' | 'receipt_digest'>>;

const BeatLaneTerminalReceiptV1InputSchema = z.object({
  run_id: NonBlankString,
  workspace_id: NonBlankString,
  package_digest: DigestSchema,
  plan_digest: DigestSchema,
  beat_index: z.number().int().nonnegative().safe(),
  lane: NonBlankString,
  status: z.enum(TERMINAL_BEAT_LANE_STATUSES_V1).optional(),
  output_digest: DigestSchema.nullable().optional(),
  receipt_digest: DigestSchema.optional(),
  contract_version: z.literal(BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1).optional(),
}).strict();

export function beatLaneTerminalReceiptDigestV1(
  receipt: Omit<BeatLaneTerminalReceiptV1, 'receipt_digest'>,
): string {
  return sha256Digest(receipt);
}

export function createBeatLaneTerminalReceiptV1(
  input: BeatLaneTerminalReceiptV1Input,
): BeatLaneTerminalReceiptV1 {
  const parsedInput = BeatLaneTerminalReceiptV1InputSchema.parse(input);
  const { receipt_digest: _ignoredReceiptDigest, ...withoutReceiptDigest } = parsedInput;
  const payload = {
    ...withoutReceiptDigest,
    status: parsedInput.status ?? 'succeeded',
    output_digest: parsedInput.output_digest ?? null,
    contract_version: parsedInput.contract_version ?? BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1,
  } as Omit<BeatLaneTerminalReceiptV1, 'receipt_digest'>;
  return BeatLaneTerminalReceiptV1Schema.parse({
    ...payload,
    receipt_digest: sha256Digest(payload),
  });
}

const CoverageContractVersionSchema = z.union([
  z.literal(BEAT_COVERAGE_CONTRACT_VERSION_V1),
  z.literal(SERIAL_FAN_IN_RECEIPT_CONTRACT_VERSION_V1),
]);

interface CoverageCandidate {
  run_id: string;
  workspace_id: string;
  package_digest: string;
  plan_digest: string;
  expected_n_beats: number;
  expected_beat_indices: number[];
  lane_receipts: BeatLaneTerminalReceiptV1[];
  required_lanes: string[];
  coverage_digest: string;
  contract_version: string;
}

function appendExpectedIndexIssue(
  coverage: CoverageCandidate,
  expected: number[],
  ctx: z.RefinementCtx,
): void {
  const invalid = coverage.expected_beat_indices.length !== expected.length
    || coverage.expected_beat_indices.some((value, index) => value !== expected[index]);
  if (invalid) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['expected_beat_indices'],
      message: 'expected_beat_indices must be exactly 0..N-1 without duplicates, holes, or out-of-range values',
    });
  }
}

function appendRequiredLaneIssues(coverage: CoverageCandidate, ctx: z.RefinementCtx): void {
  if (new Set(coverage.required_lanes).size !== coverage.required_lanes.length) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['required_lanes'], message: 'required_lanes contains duplicate lanes' });
  }
  for (const requiredLane of DEFAULT_BEAT_COVERAGE_LANES_V1) {
    if (!coverage.required_lanes.includes(requiredLane)) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['required_lanes'], message: `required_lanes must include ${requiredLane}` });
    }
  }
}

function appendReceiptBindingIssues(
  receipt: BeatLaneTerminalReceiptV1,
  coverage: CoverageCandidate,
  ctx: z.RefinementCtx,
): void {
  const bindings = [
    [receipt.run_id, coverage.run_id, 'run_id'],
    [receipt.workspace_id, coverage.workspace_id, 'workspace_id'],
    [receipt.package_digest, coverage.package_digest, 'package_digest'],
    [receipt.plan_digest, coverage.plan_digest, 'plan_digest'],
  ];
  for (const [observed, expected, name] of bindings) {
    if (observed !== expected) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['lane_receipts'], message: `lane receipt ${name} does not match coverage` });
    }
  }
  if (receipt.beat_index >= coverage.expected_n_beats) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['lane_receipts'], message: `lane receipt beat_index ${receipt.beat_index} out of range` });
  }
}

function collectReceiptKeys(coverage: CoverageCandidate, ctx: z.RefinementCtx): Set<string> {
  const seen = new Set<string>();
  for (const receipt of coverage.lane_receipts) {
    const key = `${receipt.beat_index}\u0000${receipt.lane}`;
    if (seen.has(key)) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['lane_receipts'], message: `duplicate lane receipt for ${key}` });
    }
    seen.add(key);
    appendReceiptBindingIssues(receipt, coverage, ctx);
  }
  return seen;
}

function appendMissingReceiptIssue(
  coverage: CoverageCandidate,
  expected: number[],
  seen: Set<string>,
  ctx: z.RefinementCtx,
): void {
  const expectedPairs = new Set(
    expected.flatMap((beatIndex) => coverage.required_lanes.map((lane) => `${beatIndex}\u0000${lane}`)),
  );
  const missing = [...expectedPairs].filter((key) => !seen.has(key));
  if (missing.length > 0) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['lane_receipts'], message: `missing lane receipts: ${missing.join(', ')}` });
  }
}

function appendCoverageDigestIssue(coverage: CoverageCandidate, ctx: z.RefinementCtx): void {
  const payload = { ...coverage };
  delete (payload as { coverage_digest?: string }).coverage_digest;
  payload.lane_receipts = [...payload.lane_receipts].sort((left, right) => {
    if (left.beat_index !== right.beat_index) return left.beat_index - right.beat_index;
    return left.lane < right.lane ? -1 : (left.lane > right.lane ? 1 : 0);
  });
  if (coverage.coverage_digest !== sha256Digest(payload)) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['coverage_digest'], message: 'coverage_digest does not match coverage payload' });
  }
}

export const BeatCoverageV1Schema = z.object({
  run_id: NonBlankString,
  workspace_id: NonBlankString,
  package_digest: DigestSchema,
  plan_digest: DigestSchema,
  expected_n_beats: PositiveInteger,
  expected_beat_indices: z.array(z.number().int().nonnegative().safe()),
  lane_receipts: z.array(BeatLaneTerminalReceiptV1Schema),
  required_lanes: z.array(NonBlankString).default([...DEFAULT_BEAT_COVERAGE_LANES_V1]),
  coverage_digest: DigestSchema,
  contract_version: CoverageContractVersionSchema.default(BEAT_COVERAGE_CONTRACT_VERSION_V1),
}).strict().superRefine((coverage, ctx) => {
  const expected = Array.from({ length: coverage.expected_n_beats }, (_, index) => index);
  appendExpectedIndexIssue(coverage, expected, ctx);
  appendRequiredLaneIssues(coverage, ctx);
  const seen = collectReceiptKeys(coverage, ctx);
  appendMissingReceiptIssue(coverage, expected, seen, ctx);
  appendCoverageDigestIssue(coverage, ctx);
});

export type BeatCoverageV1 = z.infer<typeof BeatCoverageV1Schema>;
export type SerialFanInReceiptV1 = BeatCoverageV1;

export type BeatCoverageV1Input = Pick<
  BeatCoverageV1,
  'run_id' | 'workspace_id' | 'package_digest' | 'plan_digest'
  | 'expected_n_beats' | 'expected_beat_indices'
> & Partial<Pick<BeatCoverageV1, 'required_lanes' | 'contract_version' | 'coverage_digest'>>
  & { lane_receipts: BeatLaneTerminalReceiptV1Input[] };

const BeatCoverageV1InputSchema = z.object({
  run_id: NonBlankString,
  workspace_id: NonBlankString,
  package_digest: DigestSchema,
  plan_digest: DigestSchema,
  expected_n_beats: PositiveInteger,
  expected_beat_indices: z.array(z.number().int().nonnegative().safe()),
  lane_receipts: z.array(BeatLaneTerminalReceiptV1InputSchema),
  required_lanes: z.array(NonBlankString).optional(),
  coverage_digest: DigestSchema.optional(),
  contract_version: CoverageContractVersionSchema.optional(),
}).strict();

export function beatCoverageDigestPayloadV1(
  coverage: Omit<BeatCoverageV1, 'coverage_digest'>,
): Omit<BeatCoverageV1, 'coverage_digest'> {
  return {
    ...coverage,
    // Completion order is not run identity; canonicalize the exact lane set.
    lane_receipts: [...coverage.lane_receipts].sort((left, right) => {
      if (left.beat_index !== right.beat_index) return left.beat_index - right.beat_index;
      return left.lane < right.lane ? -1 : (left.lane > right.lane ? 1 : 0);
    }),
  };
}

export function beatCoverageDigestV1(
  coverage: Omit<BeatCoverageV1, 'coverage_digest'>,
): string {
  return sha256Digest(beatCoverageDigestPayloadV1(coverage));
}

export function createBeatCoverageV1(input: BeatCoverageV1Input): BeatCoverageV1 {
  const parsedInput = BeatCoverageV1InputSchema.parse(input);
  const lane_receipts = parsedInput.lane_receipts.map((receipt) => {
    const payload = {
      ...receipt,
      status: receipt.status ?? 'succeeded',
      contract_version: receipt.contract_version ?? BEAT_LANE_TERMINAL_RECEIPT_CONTRACT_VERSION_V1,
    } as BeatLaneTerminalReceiptV1Input;
    return createBeatLaneTerminalReceiptV1(payload);
  });
  const {
    coverage_digest: _ignoredCoverageDigest,
    lane_receipts: _ignoredLaneReceipts,
    ...withoutCoverageDigest
  } = parsedInput;
  const payload = {
    ...withoutCoverageDigest,
    lane_receipts,
    required_lanes: parsedInput.required_lanes ?? [...DEFAULT_BEAT_COVERAGE_LANES_V1],
    contract_version: parsedInput.contract_version ?? BEAT_COVERAGE_CONTRACT_VERSION_V1,
  } as Omit<BeatCoverageV1, 'coverage_digest'>;
  return BeatCoverageV1Schema.parse({
    ...payload,
    coverage_digest: beatCoverageDigestV1(payload),
  });
}

export const SerialFanInReceiptV1Schema = BeatCoverageV1Schema;
export const BeatCoverageSchema = BeatCoverageV1Schema;
export const SerialFanInReceiptSchema = SerialFanInReceiptV1Schema;
export const createSerialFanInReceiptV1 = createBeatCoverageV1;
export const buildBeatCoverageV1 = createBeatCoverageV1;
export const buildSerialFanInReceiptV1 = createBeatCoverageV1;
